include/execution.hpp

```diff
@@ -2852,7 +2852,217 @@ namespace _P2300::execution {
+          void operator()() noexcept {
+            __stop_source_.request_stop();
+          }
+        };
+        using __on_stop = std::optional<typename stop_token_of_t<
+            env_of_t<_Receiver> &>::template callback_type<__on_stop_requested>>;
+
+        _Receiver __recvr_;
+        __on_stop __on_stop_{};
+        std::shared_ptr<__sh_state<_SenderId>> __shared_state_;
+
+      public:
+        __operation(_Receiver&& __rcvr,
+                    std::shared_ptr<__sh_state<_SenderId>> __shared_state)
+            noexcept(std::is_nothrow_move_constructible_v<_Receiver>)
+          : __operation_base{__notify}
+          , __recvr_((_Receiver&&)__rcvr)
+          , __shared_state_(move(__shared_state)) {
+        }
+        _P2300_IMMOVABLE(__operation);
+
+        static void __notify(__operation_base* __self) noexcept {
+          __operation *__op = static_cast<__operation*>(__self);
+          __op->__on_stop_.reset();
+
+          std::visit([&](const auto& __tupl) noexcept -> void {
+            std::apply([&](auto __tag, const auto&... __args) noexcept -> void {
+              __tag((_Receiver&&) __op->__recvr_, __args...);
+            }, __tupl);
+          }, __op->__shared_state_->__data_);
+        }
+
+        friend void tag_invoke(start_t, __operation& __self) noexcept {
+          __sh_state<_SenderId>* __shared_state = __self.__shared_state_.get();
+          std::atomic<void*>& __op_state1 = __shared_state->__op_state1_;
+          void* const __completion_state = static_cast<void*>(__shared_state);
+          void* const __old = __op_state1.load(std::memory_order_acquire);
+          if (__old == __completion_state) {
+            __self.__notify(&__self);
+          } else {
+              // register stop callback:
+            __self.__on_stop_.emplace(
+                get_stop_token(get_env(__self.__recvr_)),
+                __on_stop_requested{__shared_state->__stop_source_});
+            // Check if the stop_source has requested cancellation
+            if (__shared_state->__stop_source_.stop_requested()) {
+              // Stop has already been requested. Don't bother starting
+              // the child operations.
+              execution::set_stopped((_Receiver&&) __self.__recvr_);
+            } else {
+              // Otherwise, the inner source hasn't notified completion.
+              // Set this operation as the __op_state1 so it's notified.
+              __op_state1.store(&__self, std::memory_order_release);
+            }
+          }
+        }
+      };
+
+    template <class _SenderId>
+      class __sender {
+        using _Sender = __t<_SenderId>;
+        using __sh_state_ = __sh_state<_SenderId>;
+        template <class _Receiver>
+          using __operation = __operation<_SenderId, __x<remove_cvref_t<_Receiver>>>;
+
+        _Sender __sndr_;
+        std::shared_ptr<__sh_state_> __shared_state_;
+
+      public:
+        template <__decays_to<__sender> _Self, receiver _Receiver>
```

Comment:
For a single-shot sender, we must accept only rvalue instances of `__sender`.

```suggestion
        template <same_as<__sender> _Self, receiver _Receiver>
```